qemu/tests/virtio_fs_multi_vms.py

```diff
@@ -0,0 +1,96 @@
+import logging
+import os
+
+from virttest import error_context
+from provider.virtiofs_utils import run_dd_test
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Test to virtio-fs with the multiple VMs and virtiofs daemons.
+    Steps:
+        1. Create shared directories on the host.
+        2. Run virtiofs daemons on the host.
+        3. Boot guests on the host with virtiofs options.
+        4. Log into guest then mount the virtiofs targets.
+        5. Generate files on the mount points inside guests.
+        6. Compare the md5 between the host and the guest.
+        7. Compare the md5 among guests if multiple virtiofs
+           daemons share the source.
+
+    :param test: QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+    def create_dest_dir(vm, session, fs_dest):
+        error_context.context("%s: Create a destination directory %s "
+                              "inside guest." % (vm, fs_dest), logging.info)
+        session.cmd(cmd_create_dest_dir % fs_dest)
+
+    def mount_target(vm, session, fs_target, fs_dest):
+        error_context.context("%s: Mount the virtiofs target %s to %s inside guest."
+                              % (vm, fs_target, fs_dest), logging.info)
+        session.cmd(cmd_mount_target % (fs_target, fs_dest))
+
+    def umount_target(vm, session, fs_target, fs_dest):
```

Comment:
same here

Reply:
I assume that you think better put the above function into virtiofs_utils since I defined the code to a function then you think these functions should be the common api, but actually I packaged them into a function just for readability.
if my code like the following, I think you will not wanna it:
```
import logging
import os

from virttest import error_context
from provider.virtiofs_utils import run_dd_test


@error_context.context_aware
def run(test, params, env):
    """
    Test to virtio-fs with the multiple VMs and virtiofs daemons.
    Steps:
        1. Create shared directories on the host.
        2. Run virtiofs daemons on the host.
        3. Boot guests on the host with virtiofs options.
        4. Log into guest then mount the virtiofs targets.
        5. Generate files on the mount points inside guests.
        6. Compare the md5 between the host and the guest.
        7. Compare the md5 among guests if multiple virtiofs
           daemons share the source.

    :param test: QEMU test object.
    :param params: Dictionary with the test parameters.
    :param env: Dictionary with test environment.
    """
    cmd_dd = params.get('cmd_dd')
    cmd_md5 = params.get('cmd_md5')
    cmd_create_dest_dir = params.get('cmd_create_dest_dir')
    cmd_mount_target = params.get('cmd_mount_target')
    cmd_umount_target = params.get('cmd_umount_target')
    io_timeout = params.get_numeric('io_timeout')
    shared_fs_source_dir = params.get('shared_fs_source_dir')

    sessions = []
    vms = env.get_all_vms()
    for vm in vms:
        vm.verify_alive()
        sessions.append(vm.wait_for_login())

    mapping = {}
    for vm, session in zip(params.objects('vms'), sessions):
        vm_params = params.object_params(vm)
        mapping[vm] = {'session': session, 'filesystems': []}
        for fs in vm_params.objects('filesystems'):
            fs_params = vm_params.object_params(fs)
            fs_target = fs_params.get("fs_target")
            fs_dest = fs_params.get("fs_dest")
            dd_filename = 'dd_file'
            guest_data = os.path.join(fs_dest, dd_filename)
            fs_source = fs_params.get('fs_source_dir')
            mapping[vm]['filesystems'].append({'fs_target': fs_target,
                                               'fs_dest': fs_dest,
                                               'guest_data': guest_data})

            error_context.context(
                "%s: Create a destination directory %s "
                "inside guest." % (vm, fs_dest), logging.info)
            session.cmd(cmd_create_dest_dir % fs_dest)

            error_context.context(
                "%s: Mount the virtiofs target %s to %s inside guest."
                % (vm, fs_target, fs_dest), logging.info)
            session.cmd(cmd_mount_target % (fs_target, fs_dest))
            if cmd_dd:
                run_dd_test(session, cmd_dd, cmd_md5, fs_source,
                            fs_dest, dd_filename, io_timeout)
            if shared_fs_source_dir:
                continue
            error_context.context(
                "%s: Umount the viriofs target %s." % (vm, fs_target),
                logging.info)
            session.cmd(cmd_umount_target % fs_dest)

    if shared_fs_source_dir:
        error_context.context("Compare the md5 among VMs.", logging.info)
        md5_set = set()
        for vm, info in mapping.items():
            session = info['session']
            for fs in info['filesystems']:
                shared_data = fs['guest_data']
                error_context.context("%s: Get the md5 of %s." %
                                      (vm, shared_data), logging.info)
                val = session.cmd(cmd_md5 % shared_data).strip().split()[0]
                logging.info(val)
                md5_set.add(val)
                session.cmd(cmd_umount_target % fs['fs_dest'])
        if len(md5_set) != 1:
            test.fail('The md5 values are different among VMs.')
```

Reply:
And the `virtio_fs_share_data.py` if like this:
```
import logging
import os

from virttest import data_dir
from virttest import error_context
from virttest.remote import scp_to_remote

from provider.storage_benchmark import generate_instance
from provider.virtiofs_utils import run_dd_test


@error_context.context_aware
def run(test, params, env):
    """
    Test virtio-fs by sharing the data between host and guest.
    Steps:
        1. Create shared directories on the host.
        2. Run virtiofsd daemons on the host.
        3. Boot a guest on the host with virtiofs options.
        4. Log into guest then mount the virtiofs targets.
        5. Generate files or run stress on the mount points inside guest.

    :param test: QEMU test object.
    :param params: Dictionary with the test parameters.
    :param env: Dictionary with test environment.
    """
    cmd_dd = params.get('cmd_dd')
    cmd_md5 = params.get('cmd_md5')

    cmd_create_dest_dir = params.get('cmd_create_dest_dir')
    cmd_mount_target = params.get('cmd_mount_target')

    cmd_pjdfstest = params.get('cmd_pjdfstest')
    pjdfstest_pkg = params.get('pjdfstest_pkg')

    fio_options = params.get('fio_options')
    io_timeout = params.get_numeric('io_timeout')

    username = params.get('username')
    password = params.get('password')
    port = params.get('file_transfer_port')

    vm = env.get_vm(params.get("main_vm"))
    vm.verify_alive()
    session = vm.wait_for_login()
    host_addr = vm.get_address()

    for fs in params.objects("filesystems"):
        fs_params = params.object_params(fs)
        fs_target = fs_params.get("fs_target")
        fs_dest = fs_params.get("fs_dest")
        guest_data = os.path.join(fs_dest, 'fs_test')

        fs_source = fs_params.get("fs_source_dir")
        base_dir = fs_params.get("fs_source_base_dir", data_dir.get_data_dir())
        if not os.path.isabs(fs_source):
            fs_source = os.path.join(base_dir, fs_source)

        error_context.context("Create a destination directory %s "
                              "inside guest." % fs_dest, logging.info)
        session.cmd(cmd_create_dest_dir % fs_dest)
        error_context.context("Mount virtiofs target %s to %s inside guest."
                              % (fs_target, fs_dest), logging.info)
        session.cmd(cmd_mount_target % (fs_target, fs_dest))
        if cmd_dd:
            run_dd_test(session, cmd_dd, cmd_md5, fs_source,
                        fs_dest, 'dd_file', io_timeout)
        if fio_options:
            error_context.context("Run fio on %s." % fs_dest,
                                  logging.info)
            fio = generate_instance(params, vm, 'fio')
            try:
                fio.run(fio_options % guest_data, io_timeout)
            finally:
                fio.clean()
            vm.verify_dmesg()
        if cmd_pjdfstest:
            error_context.context("Run pjdfstest on %s." % fs_dest,
                                  logging.info)
            host_path = os.path.join(
                data_dir.get_deps_dir('pjdfstest'), pjdfstest_pkg)
            scp_to_remote(host_addr, port, username, password,
                          host_path, fs_dest)
            session.cmd(params.get('cmd_unpack') % fs_dest, 180)
            session.cmd(params.get('cmd_yum_deps'), 180)
            session.cmd(params.get('cmd_make') % fs_dest, io_timeout)
            session.cmd(cmd_pjdfstest % fs_dest, io_timeout)

```